wifiphisher/common/victim.py

```diff
@@ -0,0 +1,159 @@
+
+    @staticmethod
+    def get_instance():
+        if Victims.__instance is None:
+            Victims()
+        return Victims.__instance
+
+    def __init__(self):
+        if Victims.__instance is not None:
+            raise Exception("Error: Victims class is a singleton!")
+        else:
+            Victims.__instance = self
+            self.victims_dic = {}
+
+    def add_to_victim_dic(self, Victim):
+        """
+        Add new victim to the dictionary
+        """
+
+        self.victims_dic[Victim.vmac_address] = Victim
+
+    def get_print_representation(self):
+        """
+        Returns the victim dictionary in str format in order to be printed
+
+        :param self: Victims instance
+        :type self: Victims
+        :rtype str
+        """
+
+        out = ""
+        for value in self.victims_dic.values():
+            out += value.vmac_address + '\t' + value.ip_address + '\t' \
+                     + value.os + '\t' + value.vendor + '\n'
+        return out
+
+    def associate_victim_ip_to_os(self, ip_address, url):
+        """
+        receives a victims ip address and request as input, finds the
+        corresponding os by reading the initial requests file and then accesses
+        the victim dictionary and changes the os for the victim with the
+        corresponding ip address
+
+        :param self: Victims instance
+        :type self: Victims
+        :param ip_address: ip address of the victim
+        :type ip_address: str
+        :param url: request of the victim
+        :type url: str
+        """
+
+        path_to_url_file = constants.URL_TO_OS_FILE
+        with open(path_to_url_file, "r") as url_file:
+            for line in url_file:
+                line = line.split("|")
+                url_check = line[1].strip()
+                os = line[0].strip()
+                if url_check in url:
+                    for key in self.victims_dic.keys():
+                        if ip_address == self.victims_dic[key].ip_address:
+                            self.victims_dic[key].os = os
+
+    def find_devices_by_os(self, os):
+        """
+        takes as input a specific os, prints devices found with that os and
+        returns their number
+
+        :param self: Victims instance
+        :type self: Victims
+        :param os: a os (e.g android, linux)
+        :type os: str
+        :rtype int
+        """
+
+        common_os_counter = 0
+        for key in self.victims_dic:
+            if self.victims_dic[key].os == os:
+                common_os_counter += 1
+        return common_os_counter
+
+    def find_devices_by_vendor(self, vendor):
+        """
+        takes as input a specific vendor, prints devices found with that vendor
+        and returns their number
+
+        :param self: Victims instance
+        :type self: Victims
+        :param vendor: a vendor (e.g dell, intel)
+        :type vendor: str
+        :rtype int
+        """
+
+        common_vendor_counter = 0
```

Comment:
After taking a look at the code this function doesn't print anything. It just returs the number of devices right? If so the docstirng needs to be changed as well as the name.